File: python/leetcode092.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        t = ListNode(-1)
        t.next = head
        left = t
        i = 1
        while i < m:
            left = left.next
            i = i + 1
        l = left.next
        while i < n:
            l = l.next
            i = i + 1
        right = l.next
        r = left.next
        l.next = None
        pre = None
        while r:
            ne = r.next
            r.next = pre
            pre = r
            r = ne
            k = r
        left.next = pre
        k.next = right
        return head

File: leetcodec++/leetcode092.cpp
//
// Created by 黄敏辉 on 2019-07-30.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode *l,*pre, *tmp1, *tmp2;
        ListNode prehead(-1);
        prehead.next = head;
        l = &prehead;

        for (int i = 0; i < m-1; i++)
            l = l->next;
        pre = l;
        l = l->next;
        tmp1 = l;

        for (int i = 0;i < n-m;i++){
            tmp2 = l->next;
            l->next = l->next->next;
            tmp2->next = tmp1;
            pre->next = tmp2;
            tmp1 = tmp2;
        }

        return prehead.next;

    }
};